Add multiple issuer support for send-redemption-record.

dvorak42 opened the pull request:
This fixes #64.

dvorak42 commented:
@davidvancleve Main difference from existing stuff in the Chrome implementation is the movement of the alg field into the issuer specific fields of the Sec-Signature field.

davidvancleve commented on README.md:
```diff
@@ -178,14 +178,12 @@ The browser will add a new request header with the resulting signature over a co
 
 
-The canonical CBOR data (verifiable by the signature) should be computable from a request, and so does not need to be sent over the wire from the browser. The `Signed-Headers` header, and the value of `sign-request-data` should be enough to re-construct it server side, robust to things like header re-ordering, etc. In Version 3 of the protocol, the algorithm used for signing is [`ecdsa_secp256r1_ ...
+The canonical CBOR data (verifiable by the signature) should be computable from a request, and so does not need to be sent over the wire from the browser. The `Signed-Headers` header, and the value of `sign-request-data` should be enough to re-construct it server side, robust to things like header re-ordering, etc. A key for each requested issuer is included in the signatures field as a Structure ...
```
> Structured Headers list

I'd prefer leaving the details of the fields' encoding out: perhaps we could just use the word "collection" instead of the quoted phrase.

(For instance, I don't know if there's a Structured Headers notion of a "pair." It makes life a little easier if we're able to keep the examples as simple-as-possible pseudocode that shows the logical structure but doesn't try to look like the wire format. It also keeps the explainer firmly in "high-level illustration of potential functionality" territory and reduces the risk it can be misconstrued as normative.)

davidvancleve commented:
Currently, we have a 1:1 between browser versions, supported "major" Trust Tokens versions, and request-signing signature schemes.

What's the way in which we see this changing soon? Would we expect browsers to start supporting multiple "major" versions with different signature schemes, or multiple signature schemes per "major" version?

I wonder if we should hold off on the `alg` change until the time we introduce another change that alters the 1:1 major-version-to-signature-scheme relationship. It seems separate from the core of this CL, which just reflects the (existing) interface refresh that allows multiple issuers per signing operation.

dvorak42 commented:
Adding the change here means that if we add a new version that uses a different algorithm, then we don't necessarily need to break compatibility to support it, holding off on this change until that time means that we can't have backwards compatibility at that point (since V3 aware consumers wouldn't be able to consume V4 Sec-Signatures.

davidvancleve commented:
Let me rephrase to make sure I understand:
* V3 is already going to require changes from signature consumers, because it switches the signature scheme to P256
* If V4 introduces the ability for issuers to choose between multiple signature schemes, it would require we make another backwards-incompatible change to the signing operation's data structure. But if we make that change now, it might mean that signature consumers do not need code changes to support V4 signatures.

If I'm understanding right, I have a question about this. We update the "domain separator" in a backwards-incompatible manner every major version change. Doesn't this mean that signature consumers would already need to make a code change, no matter what, when the latest major version ticks up from V3 to V4?

Thanks for bearing with me as I gather the context for the change!

dvorak42 commented:
Lets say V3 and V4 of the protocol are basically the same, except V4 moves to p521 or some other signature algorithm. With this change in V3, a browser supporting both can send both V3 and V4 in the same send-redemption-record/signing thing, downstream consumers will also be compatible as long as they understand V3 at least, since they can just ignore elements they don't understand.

Without this change, trying to send a V3 Sec-Signature means you can't use V4, since you don't know whether or not the downstream consumer has updated to support the new V4 format, so you can never mix both.

If we had thought ahead, we could've done this change in V2 so we don't have to break Sec-Signature between V2 and V3.

callumdmay commented:
@dvorak42 do you think its worth adding a bit to the Version History in the Issuer Protocol README?

dvorak42 commented:
Yeah, that's probably a good idea, I'll add something to the PR for that in a bit.

davidvancleve commented:
In general, I definitely like the idea of good compat. 

I haven't spent much time thinking through this, but I don't off the bat have a great grasp of the kinds of version mismatch that this change hopes to support and whether it would involve a change from the current "every change is backwards-incompatible by definition" way we're approaching compat during the origin trial. I think talking this over online would probably do me a world of good. I'll have pretty spotty availability tomorrow, but let's find a moment to talk on Friday.

dvorak42 commented:
Added to the protocol doc changelog and added a bit about consumers needing to ignore unknown signatures.

Ignoring the compat benefits of this change, there's also the argument that the signature algorithm and the public key/signature should all be kept in the same layer to make mismatches between the algorithm and the material less likely.

davidvancleve commented:
My intuition is that a schema of

{"alg": alg,
"signatures": {key: sig, key: sig, ...}}

does not seem that error-prone to process: you store the algorithm in a
local variable and then refer to it when iterating over the dictionary. If
storing it next to each key server-side makes sense downstream, it's easy
for server-side consumers to replicate the value.


In principle, I still have this sentiment:

> I wonder if we should hold off on the `alg` change until the time we
introduce another change that alters the 1:1
major-version-to-signature-scheme relationship. It seems separate from the
core of this CL, which just reflects the (already-landed) interface refresh
that allows multiple issuers per signing operation.

Making a change to messages' layouts primarily motivated by the fact that
it might be required to support a future change that we haven't yet itself
designed (allowing multiple signatures scheme per major protocol version)
seems like it risks requiring redoing the change when we design that larger
functional unit; this mean that adapting to the first change would be
busywork for server-side consumers integrating with the feature. I think it
would be better engineering practice to first design the larger unit. Once
we do that, it would be fine to land its constituent API changes across
a couple different major versions, if that's what turns out to be easiest:
there would be a lower risk of needing to revisit the individual changes.

On Thu, Apr 15, 2021 at 5:34 AM Steven Valdez ***@***.***>
wrote:

> Added to the protocol doc changelog and added a bit about consumers
> needing to ignore unknown signatures.
>
> Ignoring the compat benefits of this change, there's also the argument
> that the signature algorithm and the public key/signature should all be
> kept in the same layer to make mismatches between the algorithm and the
> material less likely.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/WICG/trust-token-api/pull/65#issuecomment-820387140>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ANFTZDGTMG2RNM7OVV3K5ATTI3MNRANCNFSM423O5VWQ>
> .
>

dvorak42 commented:
The current implemented scheme is:

{
  signatures: [
    (issuer1, {public-key=..., signature=...}),
    (issuer2, {public-key=..., signature=...}),
    ...
  ],
  alg: ed25519
}

There's two layers of indirection between the algorithm and the actual public key material. The fact we have an algorithm field and have made a change to the sigalg already to p256 means that the meaning of the public key/signature format can change between versions, therefore we should keep those interlinked. Most other APIs that have a in-protocol algorithm tend to have them at the same level (eg CertVerify is the algorithm identifier as the prefix of the signature in TLS 1.3, signatures from WebCrypto have it as a prefix, etc).
--

On the compatibility side, making this change now means that consumers don't necessarily need to break again in the future. There is definitely a version breakage in V3, because of the swap to p256, so adding this doesn't add any overhead there. If we don't have this, any time we want to support another sigalg, we have to do a hard break between consumers (rather than allowing consumers to migrate to parsing the new format at their own pace as they pick up Trust Token N+1 support).

davidvancleve commented:
Okay, sounds good: I'm not familiar with those other APIs, so I trust your judgment if you say this will bring the shape closer in line with existing similar APIs.

davidvancleve commented:
... is there a way I can LGTM without submitting? I can't find an "approve" button.
![image](https://user-images.githubusercontent.com/55262348/115244074-6e5e8380-a0d8-11eb-8202-977c367d41b6.png)

dvorak42 commented:
@callumdmay Did the line in ISSUER_PROTOCOL Version History seem fine, or did you want that expanded out?

@davidvancleve I *think* you can go to "files changed", hit "Review changes" and do "Approve" to get the status to change, but Github UI confuses me.

davidvancleve reviewed (approved):
LGTM!

(That did the trick. Thanks!)

callumdmay commented on ISSUER_PROTOCOL.md:
```diff
@@ -486,6 +486,6 @@ struct {
 ## Version History
 
-V3 uses [`ecdsa_secp256r1_sha256`](https://tools.ietf.org/html/rfc8446#section-4.2.3) as the signing algorithm for request signing and updates the key commitment format to nest it in a version-keyed dictionary (and to move the set of keys to their own dictionary within the commitment).
+V3 uses [`ecdsa_secp256r1_sha256`](https://tools.ietf.org/html/rfc8446#section-4.2.3) as the signing algorithm for request signing and updates the key commitment format to nest it in a version-keyed dictionary (and to move the set of keys to their own dictionary within the commitment). The send redemption API is also updated to support multiple records with differing signing algorithms for future ...
```
Suggestions to be a bit more specific:

The `send-redemption-record` API is also updated to support multiple records from different issuers with differing signing algorithms for future compatibility.
